src/runtime_src/core/edge/drm/zocl/zocl_bo.c

```diff
@@ -135,12 +136,114 @@ void zocl_free_userptr_bo(struct drm_gem_object *gem_obj)
 	kfree(&zocl_bo->cma_base);
 }
 
+static struct drm_zocl_bo *
+zocl_create_cma_mem(struct drm_device *dev, size_t size)
+{
+	struct drm_gem_cma_object *cma_obj;
+	struct drm_zocl_bo *bo;
+
+	/* Allocate from CMA buffer */
+	cma_obj = drm_gem_cma_create(dev, size);
+	if (IS_ERR(cma_obj))
+		return ERR_PTR(-ENOMEM);
+
+	bo = to_zocl_bo(&cma_obj->base);
+
+	return bo;
+}
+
+static struct drm_zocl_bo *
+zocl_create_range_mem(struct drm_device *dev, size_t size, struct zocl_mem *mem)
+{
+	struct drm_zocl_dev *zdev = dev->dev_private;
+	struct drm_zocl_bo *bo = NULL;
+	struct zocl_mem *head_mem = mem;
+	int err = -ENOMEM;
+
+	bo = kzalloc(sizeof(struct drm_zocl_bo), GFP_KERNEL);
+	if (IS_ERR(bo))
+		return ERR_PTR(-ENOMEM);
+
+	err = drm_gem_object_init(dev, &bo->gem_base, size);
+	if (err) {
+		kfree(bo);
+		return ERR_PTR(err);
+	}
+
+	bo->mm_node = kzalloc(sizeof(struct drm_mm_node),
+			GFP_KERNEL);
+	if (IS_ERR(bo->mm_node)) {
+		drm_gem_object_release(&bo->gem_base);
+		kfree(bo);
+		return ERR_PTR(-ENOMEM);
+	}
+
+	mutex_lock(&zdev->mm_lock);
+
+	do { 
+		if (mem->zm_type == ZOCL_MEM_TYPE_CMA) {
+			struct drm_zocl_bo *cma_bo = zocl_create_cma_mem(dev, size);
+			if (!IS_ERR(cma_bo)) {
+				/* Get the memory in CMA memory region */
+				mutex_unlock(&zdev->mm_lock);
+				kfree(bo->mm_node);
+				drm_gem_object_release(&bo->gem_base);
+				kfree(bo);
+				cma_bo->flags |= ZOCL_BO_FLAGS_CMA;
+				return cma_bo;
+			}
+		}
+		else {
+			err = drm_mm_insert_node_generic(mem->zm_mm,
+				bo->mm_node, size, PAGE_SIZE, 0, 0);
+			if (!err) {
+				/* Got memory from this Range memory manager */
+				break;
+			}
+		}
+
+		/* No memory left to this memory manager.
+		 * Try to allocate from similer memory manger link list
+		 */
+		mem = list_entry(mem->zm_mm_list.next, typeof(*mem), zm_mm_list);
+
+	} while (&mem->zm_mm_list != &head_mem->zm_mm_list);
+	
+	if (err) {
+		DRM_ERROR("Fail to allocate BO: size %ld\n",
+				(long)size);
+		mutex_unlock(&zdev->mm_lock);
+		kfree(bo->mm_node);
+		drm_gem_object_release(&bo->gem_base);
+		kfree(bo);
+		return ERR_PTR(-ENOMEM);
+	}
+
+	mutex_unlock(&zdev->mm_lock);
+
+	/*
+	 * Set up a kernel mapping for direct BO access.
+	 * We don't have to fail BO allocation if we can
+	 * not establish the kernel mapping. We just can not
+	 * access BO directly from kernel.
+	 */
```

Comment:
Oh, some big comment! This is always appreciated...
Go on! :-)

Reply:
Thanks